src/main/java/org/opensearch/sdk/ExtensionsRunner.java

```diff
@@ -76,33 +74,67 @@ public ExtensionsRunner() throws IOException {}
     private final TransportInterceptor NOOP_TRANSPORT_INTERCEPTOR = new TransportInterceptor() {
     };
 
+    /**
+     * Instantiates a new Extensions Runner.
+     *
+     * @throws IOException if the runner failed to read settings or API.
+     */
+    public ExtensionsRunner() throws IOException {}
+
     private ExtensionSettings readExtensionSettings() throws IOException {
         File file = new File(ExtensionSettings.EXTENSION_DESCRIPTOR);
         ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
-        ExtensionSettings extensionSettings = objectMapper.readValue(file, ExtensionSettings.class);
-        return extensionSettings;
+        return objectMapper.readValue(file, ExtensionSettings.class);
+    }
+
+    private ExtensionApi readExtensionApi() throws IOException {
+        File file = new File(ExtensionApi.EXTENSION_API_DESCRIPTOR);
+        ObjectMapper objectMapper = new ObjectMapper(new YAMLFactory());
+        return objectMapper.readValue(file, ExtensionApi.class);
+    }
+
+    private void setUniqueId(String id) {
+        this.uniqueId = id;
+    }
+
+    String getUniqueId() {
+        return uniqueId;
     }
 
     private void setOpensearchNode(DiscoveryNode opensearchNode) {
         this.opensearchNode = opensearchNode;
     }
 
-    public DiscoveryNode getOpensearchNode() {
+    DiscoveryNode getOpensearchNode() {
         return opensearchNode;
     }
 
     /**
-     * Handles a plugin request from OpenSearch.  This is the first request for the transport communication and will initialize the extension and will be a part of OpenSearch bootstrap.
+     * Handles a plugin request from OpenSearch.  This is the first request for the transport communication as a part of OpenSearch bootstrap.
+     * It will initialize the opensearchNode and trigger an initial call to sendApiRequest.
      *
      * @param pluginRequest  The request to handle.
      * @return A response to OpenSearch validating that this is an extension.
      */
     PluginResponse handlePluginsRequest(PluginRequest pluginRequest) {
         logger.info("Registering Plugin Request received from OpenSearch");
-        PluginResponse pluginResponse = new PluginResponse(extensionSettings.getExtensionName());
         opensearchNode = pluginRequest.getSourceNode();
+        // Fetch the unique ID
+        for (DiscoveryExtension de : pluginRequest.getExtensions()) {
+            if (de.getName().equals(extensionSettings.getExtensionName())) {
+                setUniqueId(de.getId());
+                break;
+            }
+        }
+        if (getUniqueId() == null) {
+            throw new IllegalArgumentException(
```

Comment:
javadoc should include an `@throws` since this throws an IllegalArguementException

Reply:
@joshpalis Technically speaking, the `@throws` would only be required if the method itself declared `throws IllegalArgumentException`.  Which isn't required as it's an unchecked exception.  It only occurs in the case there's a misconfiguration where the extension name in `extensions.yml` doesn't match the extension name defined by the extension.

Which I might argue is an unnecessary requirement and one of the two names should be ignored.  However by passing the entire list of `DiscoveryExtension`s matching the names is the only reliable way to fetch the unique ID.  (Maybe we should have the extension itself delcare the uniqueId?)

I think the correct thing to do is not bother with the name check, but change `PluginRequest` (which should be renamed) to pass a single instance of the correct `DiscoveryExtension` instead of the list.   Since you'll be updating your NamedWritables PR based on this code you might consider altering that in the process.  (FYI @ryanbogan who has also been involved with this code.)

Reply:
> pass a single instance of the correct `DiscoveryExtension` instead of the list.

This implementation could probably accomplish the goals of #52.